### ISSUE TITLE
pygame on android

### ISSUE THREAD

**Issue by illume** [pygame/pygame#727]
There are various different ways to run pygame on android.

- https://github.com/kivy/python-for-android
- https://play.google.com/store/apps/details?id=ru.iiec.pydroid3
- pydroid3
- qpython
- pygame for android subset
- pygame_sdl2

**Comment by robertpfeiffer** [comment]
I got something to run
Next steps:
- [ ] submit recipe to kivy/p4a
- [ ] get bigger game working on android
- [ ] get more examples/ working on android
- [ ] write blog post
- [ ] document android button key codes
- [ ] compile portmidi for Android or find replacement
- [ ] compile libpng for Android
- [ ] compile freetype on android
- [ ] (optional) expose SCALED renderer to user
- [ ] (optional) pygame.event API for on-screen touch controls
- [ ] write example that makes use of android-specific functionality
- [ ] run test suite on Android device or in emulator
- [ ] investigate ctypes/cffi situation
- [ ] investigate pygame_sdl2 on p4a
- [ ] do OpenGL things with pyOpenGL
- [ ] port useful modules (modernGl, libnoise) and add to p4a
- [ ] refactor pygame build system?
- [ ] merge into master

**Comment by robertpfeiffer** [comment]
ModernGL uses several OpenGL 3.X and OpenGL4 features that necessitate changes before it can work on Android. CTypes doesn't work on android, but cffi does. So PyOpenGL is also out. That leaves https://github.com/FabriceSalvaire/PyOpenGLng, which I haven't tested yet.

It works well enough though.

@MightyJosip has made a reddit post https://www.reddit.com/r/pygame/comments/ft92jz/pygame_community_need_testers_for_making_the/

**Comment by illume** [comment]
The android PR was split up into other PRs (https://github.com/pygame/pygame/issues/727), the remaining part left is the setup.py and buildconfig/Setup.Android.SDL2.in

**Comment by robertpfeiffer** [comment]
cf #2337

**Comment by robertpfeiffer** [comment]
cf #2187

**Comment by ricardoV94** [comment]
Any updates on this? There seems to be no official documentation related to the new functionality

**Comment by robertpfeiffer** [comment]
> Any updates on this? There seems to be no official documentation related to the new functionality

You'd need to refer to kivy pythonforandroid documentation. It works, but not everything works.

**Comment by robertpfeiffer** [comment]
I think what's next is 
[ ] bump the pygame version to 2.0.1 on the kivy side
[ ] write a tutorial for touch controls

**Comment by Pixelsuft** [comment]
I've runned pygame on android!

1)You need to use FULLSCREEN flag.
2)Screen is auto resizing on android (surface.get_size())
3) Detect is it android:
```py
is_android = True
try:
     import android
except ImportError:
     is_android = False
```
4)Use this code for paths:
```py
import os

cur_path = os.getcwd()

def get_path(path):
     return os.path.join(cur_path, path) if is_android else path
```
5) I'm don't know, how android.show_keyboard() works, som use pygame-vkeyboard module.

**Comment by robertpfeiffer** [comment]
> You need to use FULLSCREEN flag

Yes, I made it like this on purpose. Maybe we will change this if SDL2 ever supports Samsung Dex or something like that.

> I'm don't know, how android.show_keyboard() works

use `pygame.key.start_text_input()` and `pygame.key.stop_text_input()` https://www.pygame.org/docs/ref/key.html#pygame.key.start_text_input

> to debug

`adb logcat` on the dev machine

**Comment by robertpfeiffer** [comment]
Ok, I think I am ready to write a tutorial for touch controls and touchscreen keyboard input. All the cool things are there, we just need to release a new release and point the kivy version to that.

**Comment by robertpfeiffer** [comment]
Also, I got GLContext to run, but never to do anything useful. Newer versions of modernGL do things that don't work with the OpenGL ES profile on my phone. I'll need to look into this at some point.

**Comment by lamdba** [comment]
zengl can run on android(buildozer), I changed few lines of code，delete just one opengl function (glGetBufferSubData)then it can work.( i just test sdl2+zengl)
